src/Microsoft.OData.Core/UriParser/Aggregation/EntitySetAggregateToken.cs

```diff
@@ -0,0 +1,97 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="EntitySetAggregateToken.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+#if ODATA_CLIENT
+namespace Microsoft.OData.Client.ALinq.UriParser
+#else
+namespace Microsoft.OData.UriParser.Aggregation
+#endif
+{
+    internal class EntitySetAggregateToken : AggregateExpressionToken
+    {
+        private readonly QueryToken entitySet;
+        private readonly IEnumerable<AggregateExpressionToken> expressions;
+
+        public EntitySetAggregateToken(QueryToken entitySet, IEnumerable<AggregateExpressionToken> expressions)
+        {
+            ExceptionUtils.CheckArgumentNotNull(expressions, "expressions");
+            this.expressions = expressions;
+            this.entitySet = entitySet;
+        }
+
+        public override QueryTokenKind Kind
+        {
+            get { return QueryTokenKind.EntitySetAggregateExpression; }
+        }
+
+        public IEnumerable<AggregateExpressionToken> Expressions
+        {
+            get
+            {
+                return expressions;
+            }
+        }
+
+        public QueryToken EntitySet
+        {
+            get
+            {
+                return entitySet;
+            }
+        }
+
+        /// <summary>
+        /// Merges two <see cref="EntitySetAggregateToken"/> that have the same entity set into one.
+        /// </summary>
+        /// <param name="token">The token that is going to be merged in the current one.</param>
+        /// <returns>Returns a token with the same entitySet as the caller token and the parameter and with expressions from both objects.</returns>
+        public EntitySetAggregateToken Merge(EntitySetAggregateToken token)

```

Comment:
Done. Used the static method, because it's not possible to modify the instances read only members.